rasa_nlu/extractors/crf_entity_extractor.py

```diff
@@ -0,0 +1,241 @@
+from rasa_nlu.training_data import TrainingData
+import spacy
+from spacy.gold import GoldParse
+import pycrfsuite
+import numpy as np
+from tempfile import NamedTemporaryFile
+import time
+import shutil
+
+
+class CRFEntityExtractor(Component, EntityExtractor):
+    name = "ner_crf"
+
+    context_provides = {
+        "process": ["entities"],
+    }
+
+    output_provides = ["entities"]
+
+    function_dict = {'low': lambda doc: doc[0].lower(), 'title': lambda doc: unicode(doc[0].istitle()),
+                     'word3': lambda doc: doc[0][-3:], 'word2': lambda doc: doc[0][-2:],
+                     'pos': lambda doc: doc[1], 'pos2': lambda doc: doc[1][:2],
+                     'bias': lambda doc: u'bias', 'upper': lambda doc: unicode(doc[0].isupper()),
+                     'digit': lambda doc: unicode(doc[0].isdigit())}
+
+    def __init__(self, ent_tagger=None, nlp=spacy.load('en'), crf_features=None, BILOU_flag=True):
+
+        self.ent_tagger = ent_tagger
+        self.nlp = nlp
+
+        # BILOU_flag determines whether to use BILOU tagging or not. More rigorous however requires more examples per entity
+        # rule of thumb: use only if more than 100 egs. per entity
+        self.BILOU_flag = BILOU_flag
+
+        if not crf_features:
+            # crf_features is [before, word, after] array with before, word, after holding keys about which
+            # features to use for each word, for example, 'title' in array before will have the feature "is the preceeding word in title case?"
+            self.crf_features = [['low', 'title', 'upper', 'pos', 'pos2'],
+                                 ['bias', 'low', 'word3', 'word2', 'upper', 'title', 'digit', 'pos', 'pos2'],
+                                 ['low', 'title', 'upper', 'pos', 'pos2']]
+        else:
+            self.crf_features = crf_features
+        self.config = {'crf_features': self.crf_features, 'BILOU_flag': BILOU_flag}
+
+    def train(self, training_data):
+        # type: training_data -> None
+        if training_data.num_entity_examples > 0:
+            train_data = self._convert_examples(training_data.entity_examples)
+            ent_types = [[ent["entity"] for ent in ex["entities"]] for ex in training_data.entity_examples]
+            entity_types = list(set(sum(ent_types, [])))
+
+            # convert the dataset into features
+            dataset = [self._from_json_to_crf(q) for q in train_data]
+            # train the model
+            self._train_model(dataset)
+
+    def test(self, testing_data):
+        if testing_data.num_entity_examples > 0:
+            test_data = self._convert_examples(testing_data.entity_examples)
+            ent_types = [[ent["entity"] for ent in ex["entities"]] for ex in testing_data.entity_examples]
+            entity_types = list(set(sum(ent_types, [])))
+            dataset = [self._from_json_to_crf(q) for q in test_data]
+            self._test_model(dataset)
+
+    # def process(self):
```

Comment:
No you don't need to add the entities to the spacy doc object. The only thing you need to do is return them as a json object in the same way spacy and mitie do it:
```python
return {
         "entities": self.extract_entities(text, tokens, mitie_feature_extractor)
       }
```